Dr/oauth2 improvements (#63)
ecg-icas/icas: src/source/authentication.rst

* **Authorization Server:** The authorization server is the server
  authorizing the client app to access the resources of the resource owner.
  This allows fine-grained access to resources owned by the **resource owner**.

> In our OAuth 2.0 flow we rely on the *tenant* (local market) as being the **Authentication Server**, the

Reviewer: /s/as being/providing

one that confirms the user identity through the use of some kind of credentials (usually username/password).

> The authentication server is hosting the authentication of a user account (resource owner).

Reviewer: minor/language: not sure you can say "hosting the authentication of". Maybe once more "providing" is better fitting?

>   In this context *authorization server* is the server hosting authentication

Reviewer 2: shoud this be the 'authorization and token endpoints'? as authentication happens @ the tenant

Reviewer 2: can you verify whether this is indeed correct or not?

Author: corrected

Author: good catch

  and token endpoints.

> In our Oauth2 flow we rely on the *tenant* as being the **Authentication Server**, the

Reviewer 2: Do we want to call this 'tenant' or 'local market' ?

Author: how about *tenant* (local market) :D

Author: what's the difference? the local market plays a tenant role, that's how I see it (for us)

> In our Oauth2 flow we rely on the *tenant* (local market) as being the **Authentication Server**, the

Reviewer 3: typo

```suggestion
In our OAuth 2.0 flow we rely on the *tenant* (local market) as being the **Authentication Server**, the
```

> one that confirms hte user identity through the use of some kind of credentials (usually username/password).

Reviewer 2: ```suggestion
one that confirms the user identity through the use of some kind of credentials (usually username/password).
```

> The authentication server represents a user or a service account to the **Authorization Server**.

Reviewer 2: what does 'represents' mean in this context? I don't understand this sentence.

Author: presents?owns?

> The authentication server owns a user or a service account to the **Authorization Server**.

Reviewer 2: I don't fully understand the 'account to the auth server' part; how should I read this?

Author: I meant a user or a service type of account. Removed, it just confuses :)

Author: can we consider resolved? @twessling-icas

The steps for getting an access token are as follows:

> If the user is not logge-in with the tenant:

Reviewer 2: ```suggestion
If the user is not logged-in with the tenant:
```

    GET /code?code=AUTH_CODE&state=YOUR_STATE
    Host: yoursite.com

> This is how your server gets the short-lived authorization code to subsequently get a refresh & access token.

Reviewer: It's not really "your server" but more "your client".

Author: spot on

    GET /code?code=AUTH_CODE&state=YOUR_STATE
    Host: yoursite.com

> This is how your Authorization server gets the short-lived code to further obtain a token.

Reviewer 2: your Authorization server? Your server, right?

Reviewer 2: Also, 'to subsequently get a refresh & access token'.

> This is how your server gets the short-lived code to subsequently get a refresh & access token.

Reviewer 3: suggestion:

```suggestion
This is how your server gets the short-lived authorization code to subsequently get a refresh & access token.
```

    Host: admarkt.demo.qa-mp.so
    Content-Type: application/x-www-form-urlencoded

>     -d 'grant_type=authorization_code&code=AUTH_CODE&client_id=YOUR_CLIENT_ID&client_secret=YOUR_CLIENT_SECRET&rredirect_uri=https://yoursite.com/code' https://admarkt.marktplaats.nl/accounts/oauth/token

Reviewer 4: this probably not correct if we say about post body, should be like it was before.
I see here a post body request, first headers, then data. it is not a curl cli example.

.. warning::

>     Beware of the Content-type: build it as a `multipart/form-data`.

Reviewer 2: the example above makes ut application/x-www-form-urlencoded, this is not in sync with this warning

Author: indeed we're not doing curl examples

    Beware also that you need to know the `user_id` on behalf of which the Authorization server accesses data. You can
    obtain this through `passing the proper cookie: <https://pages.github.es.ecg.tools/ecg-icas/api/doc/tenant/api/post_auth_cookie.html>`_
    from the user's browser (since it's on the same domain as the tenant).

>     Check the `tenant documentation: <https://pages.github.es.ecg.tools/ecg-icas/api/doc/tenant/api/get_config.html>`_ on how to get the `user_id` and which are the relevant cookies to configure for this.

Reviewer 2: hold up, this is external documentation referring to our internal documentation? This block shoudl not be here. 
The customer should never know the 'user_id'.

Author: then how do we get tenants to behave as sellside API partners?

Author: ah, nvm I see your point, will remove it

Author: >  You can obtain this through `passing the proper cookie:

this is not the API client actually. referred to as "You"

 * - reporting
   - Grants access to reporting endpoints only. Allows creation of detailed downloadable reports

> You can find a generalized sequence diagram of the Oauth2 flow below (click to expand).

Reviewer 3: ```suggestion
You can find a generalized sequence diagram of the OAuth 2.0 flow below (click to expand).
```
